### PR TITLE
New amount view on send tab and request specific amounts - Mobile

### PR DESCRIPTION
**Pull request by JDonadio**
image

image

image

### CODE REVIEW

**Comment by matiu** [comment]
things to check:

- [ ] BIP21 url handling (click from outside the app)
- [ ] BIP21 url handling (copy pasting to the `to` address)
- [ ] BIP21 QR scanner
- [ ] BIP70 url handling (click from outside the app)
- [ ] BIP70 url handling (copy paste)
- [ ] BIP70 QR scanner

in all the above cases, the amount field should be blocked, and `Cancel` should reset the form.

- [ ] change USD to bitcoin and viceverse in the popup and in the form (?)

**Comment by matiu** [comment]
also:
 - [ ] on desktop, browser and chrome app, the normal 'keyboard' num keys should work
 - [ ] on desktop, the Enter key should submit the amount.

File: src/js/controllers/modals/inputAmount.js [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+      var amount = evaluate(format($scope.amount));
+      if ($scope.showAlternativeAmount)
+        $scope.globalResult = '= ' + $filter('formatFiatAmount')(amount);
+      else
+        $scope.globalResult = '= ' + profileService.formatAmount(amount * unitToSatoshi, true);
+    }
+  };
+
+  function checkFontSize() {
+    if ($scope.amount && $scope.amount.length >= 13) $scope.smallFont = true;
```
Comment (isocolsky) [inline src/js/controllers/modals/inputAmount.js]: Replace magic numbers with constants defined at the start of the controller.

File: src/js/controllers/modals/inputAmount.js [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    else $scope.smallFont = false;
+  };
+
+  $scope.pushDigit = function(digit) {
+    checkFontSize();
+    if ($scope.amount && $scope.amount.length >= 19) return;
```
Comment (isocolsky) [inline src/js/controllers/modals/inputAmount.js]: See above.

File: src/js/controllers/modals/inputAmount.js [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        return val + operator;
+      } else {
+        return val.slice(0, -1) + operator;
+      }
+    };
+  };
+
+  function isOperator(val) {
+    var regex = /[\/\-\+\x\*]/;
+    var match = regex.exec(val);
```
Comment (isocolsky) [inline src/js/controllers/modals/inputAmount.js]: replace with ```return regex.test(val);``` (untested)

File: src/js/controllers/modals/inputAmount.js [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    if (match) return true;
+    return false;
+  };
+
+  function isExpression(val) {
+    var regex = /^\.?\d+(\.?\d+)?([\/\-\+\*x]\d?\.?\d+)+$/;
+    var match = regex.exec(val);
```
Comment (isocolsky) [inline src/js/controllers/modals/inputAmount.js]: same as above.

File: src/js/controllers/modals/inputAmount.js [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    if (!val) {
+      $scope.resetAmount();
+      return;
+    }
+
+    var formatedValue = format(val);
+    var result = evaluate(formatedValue);
+
+    if (lodash.isNumber(result)) {
+      if ($scope.showAlternativeAmount)
```
Comment (isocolsky) [inline src/js/controllers/modals/inputAmount.js]: This code is repeated, it should be extracted in an internal method.

File: src/js/controllers/modals/inputAmount.js [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+  function isExpression(val) {
+    var regex = /^\.?\d+(\.?\d+)?([\/\-\+\*x]\d?\.?\d+)+$/;
+    var match = regex.exec(val);
+
+    if (match) return true;
+    return false;
+  };
+
+  $scope.removeDigit = function() {
+    if ($scope.amount.toString().length == 1) {
```
Comment (isocolsky) [inline src/js/controllers/modals/inputAmount.js]: Does this test need to be here? It seems just slicing the last value and calling `processAmount()` should achieve the same result.

File: src/js/controllers/modals/inputAmount.js [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+  };
+
+  function evaluate(val) {
+    var result;
+    try {
+      result = $scope.$eval(val);
+    } catch (e) {
+      return 0;
+    }
+    if (result == 'Infinity' || lodash.isNaN(result)) return 0;
```
Comment (isocolsky) [inline src/js/controllers/modals/inputAmount.js]: using '_.isFinite()'  is cleaner.

**Comment by isocolsky** [comment]
Resulting amount not valid when thousands separator is used:
image

File: src/js/controllers/modals/inputAmount.js [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+
+    if ($scope.showAlternativeAmount) {
+      amount = fromFiat($scope.globalResult);
+      alternativeAmount = profileService.formatAmount(evaluate(format($scope.amount)) * unitToSatoshi, true);
+    } else {
+      amount = profileService.formatAmount(evaluate(format($scope.amount)) * unitToSatoshi, true);
+      alternativeAmount = toFiat(amount);
+    }
+
+    if ($scope.address) {
```
Comment (isocolsky) [inline src/js/controllers/modals/inputAmount.js]: Where is `$scope.address` defined/used?

File: src/js/controllers/modals/inputAmount.js [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+      var satToBtc = 1 / 100000000;
+      var amountSat = parseInt((amount * unitToSatoshi).toFixed(0));
+      if ($scope.unitName == 'bits') {
+        $scope.specificAmountBtc = (amountSat * satToBtc).toFixed(8);
```
Comment (isocolsky) [inline src/js/controllers/modals/inputAmount.js]: Why is this done when `unitName` == 'bits'? Also, it seems `specificAmountBtc` is not used. Is this dead code?

File: src/js/controllers/modals/inputAmount.js [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+      $scope.resetAmount();
+      return;
+    }
+
+    var formatedValue = format(val);
+    var result = evaluate(formatedValue);
+
+    if (lodash.isNumber(result)) {
+      if ($scope.showAlternativeAmount)
+        $scope.globalResult = isExpression(val) ? '= ' + $filter('formatFiatAmount')(result) : '';
```
Comment (isocolsky) [inline src/js/controllers/modals/inputAmount.js]: In this case, `$scope.globalResult` should be computed based on `$scope.amountResult` and `$scope.alternativeResult` which are already formatted.

**Comment by isocolsky** [comment]
When typing USD 9999 and pressing 'OK'  the amount will not reflect on the send screen.

**Comment by JDonadio** [comment]
Updated. Excellent review @isocolsky, thanks!

**Comment by isocolsky** [comment]
ACK! Great job!! 😄

**Comment by JDonadio** [comment]
I just added a flexible view - UI improvement

**Comment by gabrielbazan7** [comment]
ACK !

**Comment by JDonadio** [comment]
Updated. Thanks to all you guys for comments and contributions!

**Comment by cmgustavo** [comment]
ACK. Tested on iOS.

Great job!

**Pull request merged**